source/common/network/listen_socket_impl.h

```diff
@@ -244,14 +244,25 @@ class ConnectionSocketImpl : public SocketImpl, public ConnectionSocket {
 class AcceptedSocketImpl : public ConnectionSocketImpl {
 public:
   AcceptedSocketImpl(IoHandlePtr&& io_handle, const Address::InstanceConstSharedPtr& local_address,
-                     const Address::InstanceConstSharedPtr& remote_address)
-      : ConnectionSocketImpl(std::move(io_handle), local_address, remote_address) {
-    ++global_accepted_socket_count_;
+                     const Address::InstanceConstSharedPtr& remote_address,
+                     Server::ThreadLocalOverloadStateOptRef overload_state,
+                     bool track_global_cx_limit_in_overload_manager)
+      : ConnectionSocketImpl(std::move(io_handle), local_address, remote_address),
+        overload_state_(overload_state),
+        track_global_cx_limit_in_overload_manager_(track_global_cx_limit_in_overload_manager) {
+    if (!track_global_cx_limit_in_overload_manager_) {
+      ++global_accepted_socket_count_;
+    }
   }
 
   ~AcceptedSocketImpl() override {
-    ASSERT(global_accepted_socket_count_.load() > 0);
-    --global_accepted_socket_count_;
+    if (track_global_cx_limit_in_overload_manager_) {
+      overload_state_->tryDeallocateResource(
```

Comment:
I would have expected to see a pair of calls, one in the constructor to allocate, and one in the destructor to deallocate. But clearly that's not the model we're using here. Can you say more?

Reply:
@RyanTheOptimist the explanation is in comments here: https://github.com/envoyproxy/envoy/pull/28202/files/1bfae38e6f7c34c04836fe235ba37b97f359f1f3#diff-27e11486c4836b7d7f730f65c6591cc6794ffb35a4273c148fd3cb54c9ba77cfR40
Instead of checking if it is possible to allocate resource/connection and then later in the code actually allocating the resource (like the old way does), we do a single call to overload manager in tcp listener which both checks for possibility of allocation and tries to allocate resource to reduce the possibility of race conditions.

Reply:
Ah! That makes sense. Mind adding a brief comment here explaining that it was allocated in `TcpListenerImpl::rejectCxOverGlobalLimit()`?